parboiled-core/src/main/scala-3/org/parboiled2/ParserMacros.scala

```diff
@@ -42,85 +42,39 @@ private[parboiled2] trait RuleRunnable {
   }
 }
 
-object ParserMacros {
-  import scala.quoted._
-  import scala.compiletime._
-
-  def ruleImpl[I <: HList: Type, O <: HList: Type](r: Expr[Rule[I, O]])(using Quotes): Expr[Rule[I, O]] =
-    nameRuleImpl(Expr("todo"))(r)
-
-  def nameRuleImpl[I <: HList: Type, O <: HList: Type](
-      name: Expr[String]
-  )(r: Expr[Rule[I, O]])(using Quotes): Expr[Rule[I, O]] = {
-    import quotes.reflect.*
-
-    def opTreeF(rule: Expr[Rule[I, O]]): OpTree = rule match {
-      case '{
-            (${ lhs }: Rule[I, O])
-              .~((${ rhs }: Rule[I, O]))($c, $d)
-          } =>
-        Sequence(Seq(opTreeF(lhs), opTreeF(rhs)))
-      case '{ ($p: Parser).ch($c) }                 => CharMatch(p, c)
-      case '{ ($p: Parser).str($s) }                => StringMatch(p, s)
-      case '{ ($p: Parser).ignoreCase($c: Char) }   => IgnoreCaseChar(p, c)
-      case '{ ($p: Parser).ignoreCase($s: String) } => IgnoreCaseString(p, s)
-      case '{ ($p: Parser).predicate($pr) }         => CharPredicateMatch(p, pr)
-      case '{ ($p: Parser).anyOf($s) }              => AnyOf(p, s)
-      case '{ ($p: Parser).noneOf($s) }             => NoneOf(p, s)
-      case '{ ($p: Parser).ANY }                    => ANY(p)
-      case _                                        => reportError(s"Invalid rule definition: '${r.show}';", r)
-    }
-
-    val opTree: OpTree = opTreeF(r)
-
-    val parser = opTree.parser
-    '{
-      def wrapped: Boolean = ${ opTree.render(wrapped = true) }
-      val matched =
-        if ($parser.__inErrorAnalysis) wrapped
-        else ${ opTree.render(wrapped = false) }
-      if (matched) org.parboiled2.Rule.asInstanceOf[Rule[I, O]] else null
-    }
-  }
-
-  private def reportError(error: String, expr: Expr[Any])(using quotes: Quotes): Nothing = {
-    quotes.reflect.report.error(error, expr)
-    throw new Exception(error)
-  }
-
+import scala.quoted._
+class OpTreeContext(parser: Expr[Parser])(using Quotes) {
   sealed trait OpTree {
-    def parser: Expr[Parser]
```

Comment:
We don't need to capture this now anymore, it's captured once per macro and then available from the context.